src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,298 @@
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                return document.WithSyntaxRoot(
+                    // this returns the root because VB requires changing context around if statement
+                    GetRootWithInvertIfStatement(
+                        document, semanticModel, ifStatement, invertIfStyle, subsequenceSingleExitPointOpt, cancellationToken));
+            }
+
+            private void AnalyzeSubsequenceControlFlow(
+                SemanticModel semanticModel,
+                TIfStatementSyntax ifStatement,
+                out bool subsequenceEndPontIsReachable,
+                out SyntaxNode subsequenceSingleExitPointOpt)
+            {
+                subsequenceEndPontIsReachable = true;
+                subsequenceSingleExitPointOpt = null;
+
+                foreach (var range in GetSubsequentStatementRange(ifStatement))
+                {
+                    AnalyzeControlFlow(semanticModel, range, out subsequenceEndPontIsReachable, out subsequenceSingleExitPointOpt);
+                    if (!subsequenceEndPontIsReachable)
+                    {
+                        return;
+                    }
+                }
+            }
+
+            private void AnalyzeIfBodyControlFlow(
+                SemanticModel semanticModel,
+                TIfStatementSyntax ifStatement,
+                out bool ifBodyEndPointIsReachable,
+                out SyntaxNode ifBodySingleExitPointOpt)
+            {
+                AnalyzeControlFlow(
+                    semanticModel,
+                    GetIfBodyStatementRange(ifStatement),
+                    out ifBodyEndPointIsReachable,
+                    out ifBodySingleExitPointOpt);
+            }
+
+            private static void AnalyzeControlFlow(
+                SemanticModel semanticModel,
+                (SyntaxNode first, SyntaxNode last) range,
+                out bool endPointIsReachable,
+                out SyntaxNode singleExitPointOpt)
+            {
+                var flow = semanticModel.AnalyzeControlFlow(range.first, range.last);
+                endPointIsReachable = flow.EndPointIsReachable;
+                singleExitPointOpt = flow.ExitPoints.Length == 1 ? flow.ExitPoints[0] : null;
+            }
+
+            private bool SubsequentStatementsAreInTheSameBlock(TIfStatementSyntax ifStatement)
+            {
+                var (start, _) = GetSubsequentStatementRange(ifStatement).First();
+                return ifStatement.Parent == start.Parent;
+            }
+
+            protected abstract bool CanInvert(TIfStatementSyntax ifStatement);
+            protected abstract bool IsElselessIfStatement(TIfStatementSyntax ifStatement);
+
+            protected abstract int GetNearmostParentJumpStatementRawKind(TIfStatementSyntax ifStatement);
+            protected abstract bool IsEmptyStatementRange((SyntaxNode first, SyntaxNode last) range);
```

Comment:
Consider making this type parameterized on TStatementSyntax as well so we can strongly type this stuff.

Reply:
(wontfix)